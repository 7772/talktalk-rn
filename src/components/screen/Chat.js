// @flow
import React, { Component } from 'react';
import {
  SafeAreaView,
  StyleSheet,
  TouchableOpacity,
  Image,
  Text,
  View,
  Keyboard,
  Platform,
  TextInput,
  KeyboardAvoidingView,
  FlatList,
} from 'react-native';

import { Header } from 'react-navigation';

import type {
  ____ViewStyleProp_Internal as ViewStyle,
  ____TextStyleProp_Internal as TextStyle,
  ____ImageStyleProp_Internal as ImageStyle,
} from 'react-native/Libraries/StyleSheet/StyleSheetTypes';

import ChatListItem from '../shared/ChatListItem';
import EmptyListItem from '../shared/EmptyListItem';
import Button from '../shared/Button';

import { Chatroom } from '../../utils/Types';
import { IC_SMILE } from '../../utils/Icons';
import { ratio, colors } from '../../utils/Styles';
import { getString } from '../../../STRINGS';

type Styles = {
  container: ViewStyle,
  content: ViewStyle,
  viewChat: ViewStyle,
  inputChat: TextStyle,
  touchMenu: ViewStyle,
  imgMenu: ImageStyle,
  btnSend: ViewStyle,
  txtSend: TextStyle,
  viewBottom: ViewStyle,
  viewMenu: ViewStyle,
};

const styles: Styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: 'white',
    flexDirection: 'column',
    alignItems: 'center',
  },
  content: {
    flex: 1,
    justifyContent: 'center',
    alignSelf: 'stretch',

    flexDirection: 'column',
    alignItems: 'center',
  },
  viewChat: {
    width: '100%',
    borderTopWidth: 0.5,
    borderColor: 'rgb(225,225,225)',
    minHeight: 52,
    maxHeight: 100,

    flexDirection: 'row',
    justifyContent: 'flex-end',
    alignItems: 'center',
  },
  inputChat: {
    width: '80%',
    fontSize: 14,
    marginRight: 20,
    paddingLeft: 48,
    color: 'black',
  },
  touchMenu: {
    position: 'absolute',
    left: 10,
    height: '100%',
    minWidth: 20,
    justifyContent: 'center',
  },
  imgMenu: {
    width: 20,
    height: 20,
  },
  btnSend: {
    right: 8,
    backgroundColor: colors.dodgerBlue,
    borderRadius: 4,
    width: 60,
    height: 36,

    alignItems: 'center',
    justifyContent: 'center',
  },
  txtSend: {
    fontSize: 14,
    fontWeight: 'bold',
    color: 'white',
    paddingHorizontal: 5,
    paddingVertical: 10,
  },
  viewBottom: {
    position: 'absolute',
    bottom: 0,
    width: '100%',
  },
  viewMenu: {
    height: 258,
    backgroundColor: 'green',
  },
});

type Props = {
  navigation: any;
};
type State = {
  chats: any,
  showMenu: boolean,
  isLoading: boolean,
  message: string,
};

class Screen extends Component<Props, State> {
  static navigationOptions = {
    title: getString('CHAT'),
  };

  input1: any;
  input2: any;
  keyboardDidShowListener: any;
  keyboardDidHideListener: any;

  constructor(props: Props) {
    super(props);
    this.state = {
      isLoading: false,
      showMenu: false,
      message: '',
      chats: [
        {
          id: '0',
          sender: '0',
          img: null,
          message: 'hello',
          date: new Date(0),
          isPeer: true,
        },
        {
          id: '1',
          sender: '1',
          img: null,
          message: 'hello',
          date: new Date(0),
          isPeer: false,
        },
      ],
    };
  }

  componentDidMount() {
    this.keyboardDidShowListener = Keyboard.addListener('keyboardDidShow', (e) => {
      console.log('keyboardHeight', e.endCoordinates.height);
    });
  }

  componentWillUnmount() {
    this.keyboardDidShowListener.remove();
  }

  render() {
    return (
      <SafeAreaView style={styles.container}>
        <KeyboardAvoidingView
          keyboardVerticalOffset = {Header.HEIGHT + 24}
          behavior={Platform.select({
            'android': null,
            'ios': 'padding',
          })}
          style={ styles.content }
        >
          <FlatList
            style={{
              alignSelf: 'stretch',
            }}
            contentContainerStyle={
              this.state.chats.length === 0
                ? {
                  flex: 1,
                  alignItems: 'center',
                  justifyContent: 'center',
                }
                : null
            }
            keyExtractor={(item, index) => index.toString()}
            data={this.state.chats}
            renderItem={this.renderItem}
            ListEmptyComponent={<EmptyListItem>{getString('NO_CONTENT')}</EmptyListItem>}
          />
          {
            !this.state.showMenu
              ? <View
                style={styles.viewChat}
              >
                <TextInput
                  ref={(v) => this.input1 = v}
                  style={styles.inputChat}
                  multiline={true}
                  placeholder={ getString('WRITE_MESSAGE') }
                  placeholderTextColor={ colors.cloudyBlue }
                  value={this.state.message}
                  defaultValue={this.state.message}
                  onChangeText={(text) => this.setState({ message: text })}
                />
                <TouchableOpacity
                  style={styles.touchMenu}
                  onPress={this.showMenu}
                >
                  <Image style={styles.imgMenu} source={IC_SMILE}/>
                </TouchableOpacity>
                <Button
                  isLoading={this.state.isLoading}
                  onPress={this.sendChat}
                  style={styles.btnSend}
                  textStyle={styles.txtSend}
                >{getString('SEND')}</Button>
              </View>
              : null
          }
        </KeyboardAvoidingView>
        {
          this.state.showMenu
            ? <View style={styles.viewBottom}>
              <View style={styles.viewChat}>
                <TextInput
                  ref={(v) => this.input2 = v}
                  multiline={true}
                  onFocus={() => this.setState({ showMenu: false }, () => {
                    if (this.input1) {
                      this.input1.focus();
                    }
                  })}
                  style={styles.inputChat}
                  placeholder={ getString('WRITE_MESSAGE') }
                  placeholderTextColor={ colors.cloudyBlue }
                  value={this.state.message}
                  defaultValue={this.state.message}
                />
                <TouchableOpacity
                  style={styles.touchMenu}
                  onPress={this.showMenu}
                >
                  <Image style={styles.imgMenu} source={IC_SMILE}/>
                </TouchableOpacity>
                <Button
                  isLoading={this.state.isLoading}
                  onPress={this.sendChat}
                  style={styles.btnSend}
                  textStyle={styles.txtSend}
                >{getString('SEND')}</Button>
              </View>
              <View style={styles.viewMenu}/>
            </View>
            : null
        }
      </SafeAreaView>
    );
  }

  renderItem = ({ item } : any) => {
    return (
      <ChatListItem
        item={item}
      />
    );
  }

  sendChat = () => {
    console.log('sendChat');
  }

  showMenu = () => {
    console.log('showMenu');
    Keyboard.dismiss();
    this.setState({
      showMenu: !this.state.showMenu,
    });
  }

  goBack = () => {
    this.props.navigation.goBack();
  }
}

export default Screen;
