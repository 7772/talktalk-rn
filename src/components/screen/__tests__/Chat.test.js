import 'react-native';
import * as React from 'react';
import Chat from '../Chat';
// import appStore from '../../../stores/appStore';

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';
import { shallow, render } from 'enzyme';

describe('rendering test', () => {
  const wrapper = shallow(
    <Chat />,
    // <Chat store={appStore}/>,
  );

  it('renders as expected', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
