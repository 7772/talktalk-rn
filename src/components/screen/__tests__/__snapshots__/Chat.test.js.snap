// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering test renders as expected 1`] = `
<SafeAreaView
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "white",
      "flex": 1,
      "flexDirection": "column",
    }
  }
>
  <KeyboardAvoidingView
    behavior="padding"
    enabled={true}
    keyboardVerticalOffset={88}
    style={
      Object {
        "alignItems": "center",
        "alignSelf": "stretch",
        "flex": 1,
        "flexDirection": "column",
        "justifyContent": "center",
      }
    }
  >
    <FlatList
      ListEmptyComponent={
        <Shared
          style={
            Object {
              "alignItems": "center",
              "alignSelf": "stretch",
              "flex": 1,
              "justifyContent": "center",
              "width": "100%",
            }
          }
        >
          No content
        </Shared>
      }
      contentContainerStyle={null}
      data={
        Array [
          Object {
            "date": 1970-01-01T00:00:00.000Z,
            "id": "0",
            "img": null,
            "isPeer": true,
            "message": "hello",
            "sender": "0",
          },
          Object {
            "date": 1970-01-01T00:00:00.000Z,
            "id": "1",
            "img": null,
            "isPeer": false,
            "message": "hello",
            "sender": "1",
          },
        ]
      }
      disableVirtualization={false}
      horizontal={false}
      initialNumToRender={10}
      keyExtractor={[Function]}
      maxToRenderPerBatch={10}
      numColumns={1}
      onEndReachedThreshold={2}
      renderItem={[Function]}
      scrollEventThrottle={50}
      style={
        Object {
          "alignSelf": "stretch",
        }
      }
      updateCellsBatchingPeriod={50}
      windowSize={21}
    />
    <View
      style={
        Object {
          "alignItems": "center",
          "borderColor": "rgb(225,225,225)",
          "borderTopWidth": 0.5,
          "flexDirection": "row",
          "justifyContent": "flex-end",
          "maxHeight": 100,
          "minHeight": 52,
          "width": "100%",
        }
      }
    >
      <TextInput
        allowFontScaling={true}
        defaultValue=""
        multiline={true}
        onChangeText={[Function]}
        placeholder="Write message..."
        placeholderTextColor="rgb(175,194,219)"
        style={
          Object {
            "color": "black",
            "fontSize": 14,
            "marginRight": 20,
            "paddingLeft": 48,
            "width": "80%",
          }
        }
        underlineColorAndroid="transparent"
        value=""
      />
      <TouchableOpacity
        activeOpacity={0.2}
        onPress={[Function]}
        style={
          Object {
            "height": "100%",
            "justifyContent": "center",
            "left": 10,
            "minWidth": 20,
            "position": "absolute",
          }
        }
      >
        <Component
          source={
            Object {
              "testUri": "../../../assets/icons/icSmile.png",
            }
          }
          style={
            Object {
              "height": 20,
              "width": 20,
            }
          }
        />
      </TouchableOpacity>
      <Button
        activeOpacity={0.5}
        imgLeftStyle={
          Object {
            "height": 24,
            "left": 16,
            "position": "absolute",
            "width": 24,
          }
        }
        indicatorColor="white"
        isDisabled={false}
        isLoading={false}
        onPress={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "rgb(58,139,255)",
            "borderRadius": 4,
            "height": 36,
            "justifyContent": "center",
            "right": 8,
            "width": 60,
          }
        }
        textStyle={
          Object {
            "color": "white",
            "fontSize": 14,
            "fontWeight": "bold",
            "paddingHorizontal": 5,
            "paddingVertical": 10,
          }
        }
      >
        Send
      </Button>
    </View>
  </KeyboardAvoidingView>
</SafeAreaView>
`;
